### PR TITLE
Nested initialization list in struct crash fix

### PR DESCRIPTION
**Pull request by ArielG-NV**
fixes: #4489 

Problem:
1. Previously, `TopStruct var = {}` would crash if a nested member also had a initExpr assigned to `{}`. This crash is due to a particular ordering of how `{}` resolves its own type blindly setting it as an invalid type as a place holder.

Solution:
1. only set the init list type when invalid (if we checkTerm twice on the same object, this won't break the code)

Note: removed the overhaul to default initialization since it breaks Falcor and likely needs more time dedicated to it, the code history will be kept in this branch (in the mean-time it is best to fix the current regression with a more simple fix).

### CODE REVIEW

File: source/slang/slang-check-conversion.cpp [hunk 1]
```diff
@@ -512,8 +512,14 @@ namespace Slang
         UInt argCount = fromInitializerListExpr->args.getCount();
         UInt argIndex = 0;
 
-        // TODO: we should handle the special case of `{0}` as an initializer
-        // for arbitrary `struct` types here.
+        if (argCount == 0 && outToExpr && toType)
```
Comment (csyonghe) [inline source/slang/slang-check-conversion.cpp:515]: If !outToExpr, you still want to return true but not create the out expr.

File: source/slang/slang-check-impl.h [hunk 2]
```diff
@@ -2870,4 +2870,47 @@ namespace Slang
 
     RefPtr<EntryPoint> findAndValidateEntryPoint(
         FrontEndEntryPointRequest* entryPointReq);
+
+
+    [[maybe_unused]] static ConstructorDecl* _getDefaultCtor(StructDecl* structDecl)
```
Comment (csyonghe) [inline source/slang/slang-check-impl.h:2875]: Why maybe_unused is needed here?

Comment (ArielG-NV) [reply source/slang/slang-check-impl.h:2875]: slang-check-impl compiled to multiple compilation units, this function is only used across 2 compilation units

**Review by csyonghe — COMMENTED** [review]
Why do we double check the expr in the first place? Can we stop doing that?

**Comment by ArielG-NV** [comment]
> Why do we double check the expr in the first place?

This is not isolated behavior, If we have two `AssignExpr` that share the same `->right` member, we will run `checkTerm` on the `->right` member twice. This only breaks `visitInitializerListExpr` because `visitInitializerListExpr` is dependent on being called only once.

Why this crash happens now?

The behavior is showing up now due to minor ordering changes in AST building.
Previous Logic: 
1. Slang did not `ensureDecl(memberVariable, DeclCheckState::DefinitionChecked)` when checking a `structDecl`, we only auto generate `__init` for `structDecl`.
2. Naturally `ensureDecl()` on the auto generated `__init` for `structDecl`. This causes a check for `AssignExpr` (contains `initExpr`). This assigns place-holder invalid type to `initExpr` from call to `visitInitializerListExpr`
3. Slang naturally runs `ensureDecl(memberVariable, DeclCheckState::DefinitionChecked)`
4. Now we call `checkVarDeclCommon` to check the memberVariable. This runs `CheckTerm()` on our `initExpr` as well. This re-assigns place-holder invalid type from call to `visitInitializerListExpr`. We now run `coerce` to change the `initExpr` type to something valid.

New Logic:
1. Slang calls `ensureDecl(memberVariable, DeclCheckState::DefinitionChecked)` when checking a `structDecl`
2. Now we call `checkVarDeclCommon` to check the memberVariable. This runs `CheckTerm()` on our `initExpr`. This assigns place-holder invalid type from call to `visitInitializerListExpr`. We now run `coerce` to change the `initExpr` type to something valid.
3. we auto generate `__init` for `structDecl` now.
4. Naturally `ensureDecl()` on the auto generated `__init` for `structDecl`. This causes a check for `AssignExpr` (contains `initExpr`). This re-assigns place-holder invalid type to `initExpr` from call to `visitInitializerListExpr`

We were double checking the `initExpr` before. Code working was pure chance on how Slang was parsing `structDecl`s.

>  Can we stop doing that?

The solution is to not use `initExpr` directly ever (clone the AST node before using) since we will only have 1 instance of each `initExpr` in that case. This is a hack though, `visitInitializerListExpr` is fundamentally unstable with Slang AST.

**Comment by csyonghe** [comment]
Shouldn't we exit immediately if we can determine the expr is already checked?

**Comment by ArielG-NV** [comment]
> Shouldn't we exit immediately if we can determine the expr is already checked?

I don't know if this would be a problem (we may modify arg list)?

**Comment by csyonghe** [comment]
An expr should never need to be checked twice. Once it is checked it should be in its final form and not be modified elsewhere.

**Comment by kaizhangNV** [comment]
@ArielG-NV 
Does the gfx-smoke test pass?
If yes, can you enable the gfx-smoke test and re-run all the checks?

**Comment by ArielG-NV** [comment]
> Does the gfx-smoke test pass?
If yes, can you enable the gfx-smoke test and re-run all the checks?

It does pass, I can (and will) enable the test.

**Comment by kaizhangNV** [comment]
ok, can you please enable the test in this PR?

**Comment by ArielG-NV** [comment]
> ok, can you please enable the test in this PR?

we never enabled gfx-smoke since `gfx` has problems with CI

**Comment by kaizhangNV** [comment]
> > ok, can you please enable the test in this PR?
> 
> we never enabled gfx-smoke since `gfx` has problems with CI, here should be an issue open for this already.

Do you know what is the issue? And do we have issue tracking this?

**Comment by ArielG-NV** [comment]
> Do you know what is the issue? And do we have issue tracking this?


https://github.com/shader-slang/slang/issues/3176

**Review by csyonghe — APPROVED** [review]

**Pull request merged**